### PR TITLE
Disallow batch sampler with multiple IPU devices

### PR DESCRIPTION
**Pull request by otaj**
## What does this PR do?

When running on multiple IPUs, `poptorch.DataLoader` increases batch size, which is not compatible with `batch_sampler` argument at all. This was discovered during debugging of #13640 

### Does your PR introduce any breaking changes? If yes, please list them.

It is impossible to have a `BatchSampler` in while running on multiple IPUs. This never worked, however the user did not get an error message about it, but some different one, later on. Now the attempt to have explicit `BatchSampler` fails.

Note that users can still pass `batch_size` argument, which creates a default `BatchSampler` under the hood and this still works

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs) (on Slack it was)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @justusschock @awaelchli @ninginthecloud @rohitgr7 @otaj @borda @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/utilities/data.py [hunk 1]
```diff
@@ -321,8 +327,23 @@ def _dataloader_init_kwargs_resolve_sampler(
...
+            and batch_sampler.sampler == sampler
+            and dataloader.batch_size == batch_sampler.batch_size
+        ):
+            raise MisconfigurationException(
+                "It is not possible to have a batch sampler in your dataloader, when running on multiple IPU devices."
+            )
     # checking the batch sampler type is different than PyTorch default.
-    if batch_sampler is not None and (type(batch_sampler) is not BatchSampler or is_predicting):
+    if (
+        not disallow_batch_sampler
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/data.py:343]: I think the logic could be refactored like this

```python
    if batch_sampler is not None:
        if disallow_batch_sampler:
            # Check that we don't have a PyTorch default batch sampler that was instantiated in DataLoader __init__
            if not (...):
                ...
        # checking the batch sampler type is different than PyTorch default.
        elif type(batch_sampler) is not BatchSampler or is_predicting:
            ...
``

**Review by carmocca — APPROVED** [review]

File: tests/tests_pytorch/utilities/test_data.py [hunk 2]
```diff
@@ -331,6 +332,23 @@ def test_replace_dataloader_init_method(cls, args, kwargs, arg_names, dataset, c
             assert getattr(dataloader, key) == value
 
 
+def test_dataloader_disallow_batch_sampler_no_raise():
```
[the remark this answers is not in the document]
Comment (Borda) [reply tests/tests_pytorch/utilities/test_data.py:335]: :+1:

**Review by Borda — APPROVED** [review]

File: docs/source-pytorch/accelerators/ipu_basic.rst [hunk 3]
```diff
@@ -38,6 +38,9 @@ Run on multiple IPUs
 --------------------
 To use multiple IPUs set the devices to a number that is a power of 2 (i.e: 2, 4, 8, 16, ...)
 
+.. note::
+  It is not possible to use :class:`torch.utils.data.BatchSampler` in your dataloaders if you are using multiple IPUs.
```
[the remark this answers is not in the document]
Comment (otaj) [reply docs/source-pytorch/accelerators/ipu_basic.rst]: Done, somehow I didn't notice it

File: tests/tests_pytorch/utilities/test_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (otaj) [reply tests/tests_pytorch/utilities/test_data.py]: Done

File: src/pytorch_lightning/utilities/data.py (hunk 1 above)
Comment (otaj) [reply src/pytorch_lightning/utilities/data.py]: Done

File: src/pytorch_lightning/strategies/ipu.py [hunk 4]
```diff
@@ -162,6 +162,7 @@ def setup(self, trainer: "pl.Trainer") -> None:
             if self.lightning_module.trainer.enable_validation:
                 model = poptorch.inferenceModel(model=model, options=inference_opts)
                 self.poptorch_models[RunningStage.VALIDATING] = model
+                self.poptorch_models[RunningStage.SANITY_CHECKING] = model
```
[the remark this answers is not in the document]
Comment (otaj) [reply src/pytorch_lightning/strategies/ipu.py:165]: So, I'm not 100 % sure I'd call it a bug. But the piece of this PR was trying to get `self.replication_factor`, which would fail in sanity checking stage, since we already had a trainer connected, but didn't have the model saved.

Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:165]:
```suggestion
                if self.lightning_module.trainer.num_sanity_val_steps > 0:
                    self.poptorch_models[RunningStage.SANITY_CHECKING] = model
```
nit

**Review by rohitgr7 — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: src/pytorch_lightning/utilities/data.py [hunk 5]
```diff
@@ -321,27 +327,39 @@ def _dataloader_init_kwargs_resolve_sampler(
...
+            # Check that we don't have a PyTorch default batch sampler that was instantiated in DataLoader __init__
+            if not (
+                type(batch_sampler) is BatchSampler
+                and batch_sampler.sampler == sampler
+                and dataloader.batch_size == batch_sampler.batch_size
+            ):
+                raise MisconfigurationException(
+                    "It is not possible to have a batch sampler in your dataloader, "
+                    "when running on multiple IPU devices."
+                )
```
Comment (awaelchli) [inline src/pytorch_lightning/utilities/data.py:342]: @otaj I think something got missed here. There is no check anywhere that we are dealing with IPU here. 
See also the IPUStrategy where we convert to the poptorch Dataloader wrapper.

Comment (otaj) [reply src/pytorch_lightning/utilities/data.py:342]: It's a bit hidden, but the argument `disallow_batch_sampler` is by default `False` and only place where we are passing any value there is in https://github.com/Lightning-AI/lightning/pull/13854/files#diff-1de8258ec09ab5d1d13840e70c64b67e9dfc098de0eba1e9f6020ed191381cc8R234

Comment (awaelchli) [reply src/pytorch_lightning/utilities/data.py:342]: Okay thanks. It makes sense. 
I was blindly looking for IPU references and couldn't spot this one.
